File: Medium/2793-count-the-number-of-complete-components/count-the-number-of-complete-components.py
class Solution:
    def solve(self, node, adj, cnt, visited, n):
        visited[node]=True
        
        cnt[0] += 1
        cnt[1] += len(adj[node])

        for it in adj[node]:
            if not visited[it]:
                self.solve(it, adj, cnt, visited, n)

    def countCompleteComponents(self, n: int, e: List[List[int]]) -> int:
        adj=[[] for _ in range(n)]

        for u, v, in e:
            adj[u].append(v)
            adj[v].append(u)

        count=0
        visited=[False]*n
        for i in range(n):
            if not visited[i]:
                cnt=[0, 0]
                self.solve(i, adj, cnt, visited, n)
                
                ve=cnt[0]
                te=cnt[1]//2;
                if te == ve*(ve-1)//2:
                    count += 1

        return count

File: 2793-count-the-number-of-complete-components/count-the-number-of-complete-components.java
class Solution{
    void solve(int node, List<List<Integer>> adj, AtomicInteger v, AtomicInteger t, boolean[] visited, int n){
        visited[node]=true;
        v.incrementAndGet();
        t.addAndGet(adj.get(node).size());

        for(int it: adj.get(node)){
            if(!visited[it]){
                solve(it, adj, v, t, visited, n);
            }
        }
    }

    public int countCompleteComponents(int n, int[][] e){
        List<List<Integer>> adj=new ArrayList<>();
        for(int i=0; i<n; i++){
            adj.add(new ArrayList<>());
        }

        for(int[] it: e){
            adj.get(it[0]).add(it[1]);
            adj.get(it[1]).add(it[0]);
        }

        int count=0;
        boolean[] visited=new boolean[n];
        for(int i=0; i<n; i++){
            if(!visited[i]){
                AtomicInteger v=new AtomicInteger(0);
                AtomicInteger t=new AtomicInteger(0);
                solve(i, adj, v, t, visited, n);
                
                int ve=v.get();
                int te=t.get()/2;
                if(te == (ve*(ve-1))/2)
                    count++;
            }
        }

        return count;

    }
}